### ISSUE TITLE
Constructing dicts with tuples confuses pylance

### ISSUE THREAD

**Issue by A-UNDERSCORE-D** [microsoft/pyright#889]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pylance appears to believe that a Dict (or OrderedDict) is not a valid `Mapping` when created using tuples:
```
Expression of type "OrderedDict[Literal['test', 'test2'], Literal[1, 2]]" cannot be assigned to declared type "Mapping[str, Any]"
  TypeVar "_KT" is invariant
    Type "Literal['test', 'test2']" cannot be assigned to type "str"Pylance (reportGeneralTypeIssues)
```

**To Reproduce**
Use below code with pylance

**Expected behavior**
No type error reported as the types are correctly defined

**Code**
```py
from typing import Mapping, Any
from collections import OrderedDict


x: Mapping[str, Any]

x = OrderedDict([('test', 1), ('test2', 2)])   # breaks
x = {"test": 1, "test2": 2}                    # works
x = dict([('test1', 1), ('test2', 2)])         # breaks

```

**VS Code extension or command-line**
vscode via pylance: v2020.7.3

**Comment by A-UNDERSCORE-D** [comment]
As an additional issue with this, assigning to keys in mappings made this way also show errors if the key is not in the original set of `Literal`s

**Comment by erictraut** [comment]
Thanks for the bug report. The first issue you reported be fixed in the next version of Pyright.

The second issue (your second post) is expected behavior if the case where the inferred type includes literal type arguments. For more details about how Pyright handles type inference of tuples and lists, refer to [this documentation](https://github.com/microsoft/pyright/blob/master/docs/type-inference.md#type-inference).

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.58, which I just published. It will also be included in the next version of Pylance.

**Comment by A-UNDERSCORE-D** [comment]
Fix has appeared in pylance and I have tested it -- It does work for type annotations on variables, however, a method accepting a `Union[Mapping[str, Any], List[Mapping[str, Any]]]` that is called with an OrderedDict appears to still show an error. Interestingly this _only_ happens when the Union is involved (it works fine in either case otherwise). 

Additionally, this works fine if `AnyStr` is used, which is strange as that particular alias is of `str, bytes`, with no mention of Literals (though this is py3.7 I am testing under -- PyLance version 2020.7.4). It would make sense to me that literal strings can be assigned to parameters of the type str (as in `x: str = 'test'`)

example code:
```py
def test(x: Union[Mapping[str, Any], List[Mapping[str, Any]]]):
    print(x)


test(OrderedDict([
    ('test1', 1),
    ('test2', 2),
]))

```

Error:
```
Argument of type "OrderedDict[Literal['test1', 'test2'], Literal[1, 2]]" cannot be assigned to parameter "x" of type "Mapping[str, Any] | List[Mapping[str, Any]]" in function "test"
  "OrderedDict[Literal['test1', 'test2'], Literal[1, 2]]" is incompatible with "List[Mapping[str, Any]]"
    TypeVar "_KT" is invariant
      Type "Literal['test1', 'test2']" cannot be assigned to type "str"
```

**Comment by erictraut** [comment]
That's correct. Bidirectional inference doesn't support unions for the "expected type". You'll need to define intermediate variables with their own type annotations in this case.
